### ISSUE TITLE
LXQt-runner calculator not working anymore

### ISSUE THREAD

**Issue by stefonarch** [lxqt/lxqt#1672]
On both my systems it's some time that it doesn't work anymore, nothing happens typing in the runner
`2+4=`

It could be a qt related problem, qt5-base upgraded 20 gen 2019 09:16:15 CET

##### Expected Behavior

It should display the result

##### Current Behavior

It doesn't.

 
##### System Information

* Distribution & Version: git packages, arch linux | arch packages (32-bit)
* Qt Version: 5.12.0-3 and 5.12.0-2 (32-bit) with LXQt 0.13

**Comment by tsujan** [comment]
With Qt 5.12.1 (in Manjaro and so, from Arch), Runner's calculator works here without problem -- actually, it was the first time I used it ;)

So, it may be really an upgrade problem in your case.

**Comment by agaida** [comment]
erm - dunno how to say - don't work for me even.

**Comment by tsujan** [comment]
Then, we can conclude that it works for those whose don't use it -- a very effective thinking style in programming ;)

**Comment by agaida** [comment]
a new incarnation of a modified Schrödingers Bug

**Comment by tsujan** [comment]
I haven't read Runner's code but I think it's about timing.

**Comment by agaida** [comment]
erm - don't know if i should laugh or cry - just built a new siduction lxqt iso and run it in a remote KVM - can someone guess the outcome?
![](https://i.imgur.com/Obvgl9O.png)
And it seems not to be a modified Schrödingers, but a Heisenbug

**Comment by tsujan** [comment]
> And it seems not to be a modified Schrödingers, but a Heisenbug

Have seen lots of them in libfm-qt and all of them where about timing; once timing got fixed, they were eliminated.

**Comment by Chiitoo** [comment]
Seems to work on Gentoo, with Qt 5.13.0 in the making as well.

I know no one asked about that, but I had to chime in because of the “works for those who don't use it” comment.  I didn't even know this functionality exists before now, so thanks for making me aware!  :]

**Comment by tsujan** [comment]
@agaida @stefonarch One question: Does it *never* work for you or sometimes it does, sometimes it doesn't?

**Comment by agaida** [comment]
I know it was working somewhen - but now it never work.

**Comment by tsujan** [comment]
Then, it may not be related to timing.

The code is new to me; someone who's familiar with it should find the needle in the haystack.

**Comment by stefonarch** [comment]
It's some week, and it's never working here too. I had a disk failure this morning (pcie mini disk, never again)  and reinstalled arch from scratch, same issue.

**Comment by tsujan** [comment]
@palinek might help us if the issue happens to him too. I can't reproduce it to debug.

**Comment by palinek** [comment]
No, I can't reproduce. Here it works with the equal sign either on begin or on end...

**Comment by agaida** [comment]
a typical heisenbug

**Comment by tsujan** [comment]
I think @palinek uses Debian, like @agaida. So, it can't be about versions. Just for the record, here `muparser` (used for Runner's math) is 2.2.6.

**Comment by palinek** [comment]
Here the version is also 2.2.6.1:
```
libmuparser2v5/testing,now 2.2.6.1+dfsg-1 amd64 [installed,automatic]
  fast mathematical expressions parser library (runtime)
```

**Comment by stefonarch** [comment]
I forgot to mention: 32bit version arch uses still LXQt 0.13 
Just tested Lubuntu 18.10, it works there... qt 5.11.1

**Comment by tsujan** [comment]
But @agaida definitely has the latest LXQt.

BTW, can't git lxqt from AUR be compiled for 32-bit? I think a small change in PKGBUILDs should work.

**Comment by agaida** [comment]
it can - i386 should not be a problem, but a different project in Arch

**Comment by tsujan** [comment]
> maybe it's a fucked up nvidia driver or a to new kernel

I doubt that a graphic driver or kernel could cause it. Last night, I took a quick look at the code and found nothing wrong -- although I might have missed some things.

**Comment by agaida** [comment]
ok. there is something else fucked up - i take the time and my not really existent C++ knowledge and changed the relevant code a bit:

``` diff
diff --git a/providers.cpp b/providers.cpp
index e80b988..c9a47be 100644
--- a/providers.cpp
+++ b/providers.cpp
@@ -848,6 +848,7 @@ bool MathItem::run() const
  ************************************************/
 bool MathItem::compare(const QRegExp &regExp) const
 {
+    qDebug() << "in MathItem ";
     QString s = regExp.pattern().trimmed();
 
     bool is_math = false;
@@ -855,11 +856,13 @@ bool MathItem::compare(const QRegExp &regExp) const
     {
         is_math = true;
         s.remove(0, 1);
+        qDebug() << "is MathItem ";
     }
     if (s.endsWith(QLatin1Char('=')))
     {
         is_math = true;
         s.chop(1);
+        qDebug() << "is MathItem ";
     }
 
     if (is_math)
@@ -880,7 +883,8 @@ bool MathItem::compare(const QRegExp &regExp) const
                     break;
                 } catch (const mu::Parser::exception_type & e)
                 {
-                    //don't do anything, return false -> no result will be showed
+                   qDebug() << "MuParser exception "; 
+                   //don't do anything, return false -> no result will be showed
                 }
             }
         }
```
output:

`´`
~ % lxqt-runner
 (0x7ffd4e9a6f90) Debug: systemd: "CanHibernate" = "yes"
 (0x7ffd4e9a6f90) Debug: systemd: "CanSuspend" = "yes"
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: is MathItem 
 (0x7ffd4e9a6f90) Debug: MuParser exception 
 (0x7ffd4e9a6f90) Debug: MuParser exception 
 (0x7ffd4e9a6f90) Debug: MuParser exception 
 (0x7ffd4e9a6f90) Debug: MuParser exception 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: is MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
 (0x7ffd4e9a6f90) Debug: in MathItem 
```

**Comment by tsujan** [comment]
@agaida Yes, that's a good way of finding problems. "MuParser exception" in the output can be helpful. Understandably, I could have it only if I type an invalid string like:
```
2*a=
````
In that case, the "result" *will be* shown as:
```
2=2
```
Are you sure that you typed a *valid* string (like `2*2=`) to see it?

**Comment by tsujan** [comment]
If the answer to my last question is "yes", then something may be wrong in your localization. Please check that!

**Comment by agaida** [comment]
% LANG=C lxqt-runner
 (0x7ffc27ebda30) Debug: systemd: "CanHibernate" = "yes"
 (0x7ffc27ebda30) Debug: systemd: "CanSuspend" = "yes"
 (0x7ffc27ebda30) Debug: in MathItem 
 (0x7ffc27ebda30) Debug: in MathItem 
 (0x7ffc27ebda30) Debug: in MathItem 
 (0x7ffc27ebda30) Debug: in MathItem 
 (0x7ffc27ebda30) Debug: in MathItem 
 (0x7ffc27ebda30) Debug: in MathItem 
 (0x7ffc27ebda30) Debug: in MathItem 
 (0x7ffc27ebda30) Debug: is MathItem 
 (0x7ffc27ebda30) Debug: MuParser exception 
 (0x7ffc27ebda30) Debug: MuParser exception 
 (0x7ffc27ebda30) Debug: MuParser exception 
 (0x7ffc27ebda30) Debug: in MathItem 
 (0x7ffc27ebda30) Debug: is MathItem 

1+2=

**Comment by palinek** [comment]
@agaida try with this debug also:
```diff
diff --git a/providers.cpp b/providers.cpp
index e80b988..dcb0b42 100644
--- a/providers.cpp
+++ b/providers.cpp
@@ -880,6 +880,7 @@ bool MathItem::compare(const QRegExp &regExp) const
                     break;
                 } catch (const mu::Parser::exception_type & e)
                 {
+                    qDebug().nospace() << "exception expr: " << e.GetExpr().c_str() << ", msg: " << e.GetMsg().c_str();
                     //don't do anything, return false -> no result will be showed
                 }
             }
```

**Comment by agaida** [comment]
ok, some  tests later - lxqt-runner's calculator works fine if built against the muparser from stable (recompiled for sid) - will apply palinek's line now to the debugging code

**Comment by agaida** [comment]
```
 (0x7ffdf1f16700) Debug: is MathItem 
 (0x7ffdf1f16700) Debug: exception expr: , msg: Decimal separator is identic to function argument separator.
 (0x7ffdf1f16700) Debug: exception expr: , msg: Decimal separator is identic to function argument separator.
 (0x7ffdf1f16700) Debug: exception expr: , msg: Decimal separator is identic to function argument separator.
```

**Comment by tsujan** [comment]
Maybe related to http://beltoforion.de/article.php?a=muparser&p=localization

**Comment by agaida** [comment]
it is

**Comment by tsujan** [comment]
I don't know how that can be translated into [QLocale](https://doc.qt.io/qt-5/qlocale.html) methods.

**Comment by stefonarch** [comment]
Just tested the muparser-2.2.5-3 in arch, calc works again too.

**Comment by tsujan** [comment]
> Just tested the muparser-2.2.5-3 in arch, calc works again too.

With the German localization?

**Comment by tsujan** [comment]
If it's our fault -- in other words, if we haven't considered how muparser works with localization -- then it should happen with all versions and with the German localization. Otherwise, it was a muparser bug fixed in later versions.

**Comment by agaida** [comment]
erm - it might be about a special LC_$foo. And our handling of might be not the best:

```
class MathItem::Parser : public mu::Parser
{
public:
    static void initLocale()
    {
        try
        {
            // use the system's locale instead of the "C"
            s_locale = std::locale{""};
        } catch (const std::runtime_error & e)
        {
            qWarning().noquote() << "Unable to set locale for Math, " << e.what();
        }
    }
};
```

**Comment by agaida** [comment]
![](https://i.imgur.com/d1ypPSI.png) - with locale ("C") - a bit inconsitent, but even worse ![](https://i.imgur.com/O4mIJzb.png)

**Comment by agaida** [comment]
``` 
(0x7ffe77aa2960) Debug: exception expr: , msg: Decimal separator is identic to function argument separator.
``` 
Reading this again will lead us to a simple fix :D

**Comment by tsujan** [comment]
OK, I'm convinced that the pproblem is in lxqt-runner. I haven't grasped muparser+localization though.

**Comment by tsujan** [comment]
On second thought, the problem may be related to the line-edit and not to how muparser is handled because the result is shown correctly when `4,2` is typed as `4.2` inside the line-edit. kcalc doesn't have this issue but speedcrunch does.

**Comment by agaida** [comment]
if we can set the Argument Separator to something crazy like a pipe, double cross or anything else the problem will be gone.
```
src/muParserError.cpp:    m_vErrMsg[ecLOCALE]                 = _T("Decimal separator is identic to function argument separator.");

 src/muParserBase.cpp 
    // Check locale compatibility
    if (m_pTokenReader->GetArgSep()==std::use_facet<numpunct<char_type> >(s_locale).decimal_point())
      Error(ecLOCALE);

```

**Comment by stefonarch** [comment]
> > Just tested the muparser-2.2.5-3 in arch, calc works again too.
> 
> With the German localization?

italian, can test german if it's necessary

**Comment by agaida** [comment]
The relevant change in muparser:
![](https://i.imgur.com/hwD7Ev1.png)

**Comment by tsujan** [comment]
> the only thing about is: Ey, i don't know C++

I know how locale is handled by Qt but have no idea about muparser; need to get familiar with muparser later.

**Comment by agaida** [comment]
the "only" problem is that the arguments separator is the same as the decimal separator in some languages - in German it is ',' and ',' :D

**Comment by agaida** [comment]
@stefonarch - it would be nice if you could test the patch

**Comment by agaida** [comment]
1[](https://i.imgur.com/waSlpXk.png)

**Comment by tsujan** [comment]
@agaida By using what you did, I added these lines to `MathItem::MathItem()` instead:
```c++
    QLocale l = QLocale::system();
    mParser->SetArgSep(';');
    mParser->SetDecSep(l.decimalPoint().toLatin1());
    mParser->SetThousandsSep(l.groupSeparator().toLatin1());
```
And I got:

![decimal](https://user-images.githubusercontent.com/981076/52836787-527a8400-3101-11e9-91fd-99012c37e516.png)

So far, so good. However, the problem is that when the user types `1.2` instead of `1,2` *with German*, he gets a wrong result. That doesn't happen with kcalc: in kcalc, when the user types `1.2` with German, he sees `1,2` instead.

My point:
In a complete solution, not only the output but also the input should have localized decimal and thousand separators.

Moreover, as for the argument separator, we should choose a character that isn't used in any language for decimals or thousands. There's no guarantee that `;` is so.

**Comment by tsujan** [comment]
OK, kcalc does it wrong: it always translates `.` to `,` with German while `.` can be the thousand separator. For example, if I type `1.000` in kcalc, I'll get `1,000`.

So, IMO, we could expect that the user should type his/her localized separators when he uses localization, so that the above-mentioned code works.

The problems that remain:

(1) Choosing the correct argument separator. I couldn't find anything like it in `QLocale`.
(2) `QLocale` also respects the localized exponential character, negative/positive signs and percent character. I have no idea how muparser treats them.

**Comment by stefonarch** [comment]
> @stefonarch - it would be nice if you could test the patch

It works when input is correct but as mentioned above if I use . instead of , in italian I get a mess, example `2.2*2=2`

**Comment by tsujan** [comment]
> if I use . instead of , in italian I get a mess

If Italian uses `,` as the decimal separator, you *should* get a mess; shouldn't you?

As I mentioned above, kcalc does it wrong. The user should input in the localized way when he uses localization.

**Comment by palinek** [comment]
How about lxqt/lxqt-runner#118 ?

**Comment by tsujan** [comment]
> How about lxqt/lxqt-runner#118 ?

![bad](https://user-images.githubusercontent.com/981076/52847384-430b3300-3121-11e9-8caa-d602309eccb5.png)

**Comment by agaida** [comment]
my point was only the minimal approach - we should take care only of the obvious things. Sanitize all inputs would be a kind of overkill i guess.

**Comment by tsujan** [comment]
@agaida @stefonarch If you're satisfied with a non-localized input, then just setting the argument separator -- as @agaida proposed -- is the best solution. I don't think anyone uses Runner for complex math.

**Comment by agaida** [comment]
It is fully localized - with some flaws - standard german settings ( a point separates thousands, colon decimals)
![](https://i.imgur.com/CHsbCEh.png)

However - it will not work right as mentioned if one use false separators for his language.

**Comment by tsujan** [comment]
> However - it will not work right as mentioned if one...

That can be tolerated, IMO.

**Comment by agaida** [comment]
fixed in lxqt/lxqt-runner/pull/118
